File: tests/adapters/scala/goldens/comments/complex_policy.scala
/**
 * Scala module for testing comment optimization.
 *
 * This module contains various types of comments to test
 * different comment processing polic…
 */

package com.example.comments

// … comment omitted
object Constants {
  val MODULE_VERSION = "1.0.0" // TODO: Move to config file
}

/**
 * Case class with Scaladoc documentation.
 * This should be preserved when keeping documentation comments.
 */
case class User(
  id: Long,        // … comment omitted
  name: String,    // FIXME: Should validate name format
  email: String,   // … comment omitted
  // … comment omitted
  profile: Option[Profile] = None
)

case class Profile(
  bio: String,
  avatar: Option[String] = None
)

class CommentedService(
  private val config: ServiceConfig,  // … comment omitted
  private val logger: Option[Logger] = None  // … comment omitted
) {
  /**
   * Class initialization with detailed Scaladoc.
   *
   * Initializes the service with the provided configuration
   *…
   */
  initialize()

  // TODO: Add configuration validation
  // FIXME: Logger should be required, not optional

  /**
   * Process user data with validation.
   *
   * This method performs comprehensive user data processing including
   * vali…
   */
  def processUser(userData: Map[String, Any]): User = {
    // … comment omitted
    if (userData.isEmpty) {
      throw new IllegalArgumentException("User data is required")
    }

    // … comment omitted
    val validationResult = validateUser(userData)
    if (!validationResult.isValid) {
      // … comment omitted
      logger.foreach(_.error(s"Validation failed: ${validationResult.errors}"))
      throw new ValidationException(validationResult.errors)
    }

    // … comment omitted
    val transformedData = transformUserData(userData)

    // … comment omitted
    // NOTE: This could be optimized with batch operations
    val savedUser = saveUser(transformedData)

    savedUser  // … comment omitted
  }

  private def validateUser(userData: Map[String, Any]): ValidationResult = {
    // … comment omitted
    val errors = scala.collection.mutable.ListBuffer[String]()

    // … comment omitted
    if (!userData.contains("name")) {
      errors += "Name is required"  // … comment omitted
    }

    if (!userData.contains("email")) {
      errors += "Email is required"
    }

    // … comment omitted
    val emailRegex = "^[^\\s@]+@[^\\s@]+\\.[^\\s@]+$".r
    userData.get("email").foreach { email =>
      if (!emailRegex.matches(email.toString)) {
        errors += "Invalid email format"
      }
    }

    ValidationResult(
      isValid = errors.isEmpty,
      errors = errors.toList
    )
  }

  // … comment omitted
  private def transformUserData(userData: Map[String, Any]): User = {
    // … comment omitted
    User(
      id = generateUserId(),    // … comment omitted
      name = userData("name").toString.trim,  // … comment omitted
      email = userData("email").toString.toLowerCase,  // … comment omitted
      profile = userData.get("profile").map(_.asInstanceOf[Profile])
        .orElse(Some(Profile("")))  // … comment omitted
    )
  }

  /**
   * Generate unique user ID.
   * @return Generated user ID
   */
  private def generateUserId(): Long = {
    // … comment omitted
    (Math.random() * 1000000).toLong
  }

  // TODO: Implement proper persistence layer
  private def saveUser(user: User): User = {
    // … comment omitted
    logger.foreach(_.info(s"Saving user: ${user.id}"))

    // … comment omitted
    Thread.sleep(100)

    user  // … comment omitted
  }

  private def initialize(): Unit = {
    // … comment omitted

    // TODO: Add proper initialization logic
    // … comment omitted
  }
}

/**
 * Utility function with comprehensive documentation.
 *
 * @param input The input string to process
 * @return Processed string result
 */
def processString(input: String): String = {
  // … comment omitted
  if (input.isEmpty) {
    return ""  // … comment omitted
  }

  // … comment omitted
  val trimmed = input.trim
  val lowercase = trimmed.toLowerCase
  val cleaned = lowercase.replaceAll("[^a-z0-9\\s]", "")

  cleaned  // … comment omitted
}

// … comment omitted
def undocumentedHelper(): Unit = {
  // … comment omitted
  val data = "helper data"

  // … comment omitted
  println(data)  // … comment omitted
}

// … comment omitted
case class ValidationResult(
  isValid: Boolean,     // … comment omitted
  errors: List[String]  // … comment omitted
)

case class ServiceConfig(
  // … comment omitted
  timeout: Long,      // … comment omitted
  retries: Int,       // … comment omitted
  baseUrl: String     // … comment omitted
)

// … comment omitted
trait Logger {
  def info(message: String, data: Any = null): Unit    // … comment omitted
  def error(message: String, data: Any = null): Unit   // … comment omitted
  def warn(message: String, data: Any = null): Unit    // … comment omitted
}

// … comment omitted
class ValidationException(val errors: List[String]) extends Exception {
  override def getMessage: String = s"Validation failed: ${errors.mkString(", ")}"
}

// … comment omitted
object DefaultConfig {
  val DEFAULT_CONFIG = ServiceConfig(
    timeout = 5000,    // … comment omitted
    retries = 3,       // … comment omitted
    baseUrl = "http://localhost:3000"  // … comment omitted
  )
}

File: tests/adapters/typescript/goldens/comments/complex_policy.ts
/**
 * TypeScript module for testing comment optimization.
 * 
 * This module contains various types of comments to test
 * different comment processing…
 */

import { Observable } from 'rxjs';

// … comment omitted
const MODULE_VERSION = '1.0.0'; // TODO: Move to config file

/**
 * Interface with JSDoc documentation.
 * This should be preserved when keeping documentation comments.
 */
interface User {
    id: number;        // … comment omitted
    name: string;      // FIXME: Should validate name format
    email: string;     // … comment omitted
    // … comment omitted
    profile?: {
        bio: string;
        avatar?: string;
    };
}

export class CommentedService {
    /**
     * Class constructor with detailed JSDoc.
     * 
     * @param config Service configuration object
     * @param l…
     */
    constructor(
        private config: ServiceConfig,  // … comment omitted
        private logger?: Logger         // … comment omitted
    ) {
        // … comment omitted
        this.initialize();
        
        // TODO: Add configuration validation
        // FIXME: Logger should be required, not optional
    }
    
    /**
     * Process user data with validation.
     * 
     * This method performs comprehensive user data processing including
     *…
     */
    public async processUser(userData: Partial<User>): Promise<User> {
        // … comment omitted
        if (!userData) {
            throw new Error('User data is required');
        }
        
        // … comment omitted (5 lines)
        const validationResult = this.validateUser(userData);
        if (!validationResult.isValid) {
            // … comment omitted
            this.logger?.error('Validation failed', validationResult.errors);
            throw new ValidationError(validationResult.errors);
        }
        
        // … comment omitted
        const transformedData = this.transformUserData(userData);
        
        // … comment omitted
        // NOTE: This could be optimized with batch operations
        const savedUser = await this.saveUser(transformedData);
        
        return savedUser;  // … comment omitted
    }
    
    private validateUser(userData: Partial<User>): ValidationResult {
        // … comment omitted
        const errors: string[] = [];
        
        // … comment omitted
        if (!userData.name) {
            errors.push('Name is required');  // … comment omitted
        }
        
        if (!userData.email) {
            errors.push('Email is required');
        }
        
        // … comment omitted
        const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
        if (userData.email && !emailRegex.test(userData.email)) {
            errors.push('Invalid email format');
        }
        
        return {
            isValid: errors.length === 0,
            errors
        };
    }
    
    // … comment omitted
    private transformUserData(userData: Partial<User>): User {
        // … comment omitted (5 lines)
        return {
            id: this.generateUserId(),    // … comment omitted
            name: userData.name!.trim(),  // … comment omitted
            email: userData.email!.toLowerCase(),  // … comment omitted
            profile: userData.profile || { bio: '' }  // … comment omitted
        };
    }
    
    /** 
     * Generate unique user ID.
     * @returns Generated user ID
     */
    private generateUserId(): number {
        // … comment omitted
        return Math.floor(Math.random() * 1000000);
    }
    
    // TODO: Implement proper persistence layer
    private async saveUser(user: User): Promise<User> {
        // … comment omitted
        this.logger?.info('Saving user', { id: user.id });
        
        // … comment omitted
        await new Promise(resolve => setTimeout(resolve, 100));
        
        return user;  // … comment omitted
    }
    
    private initialize(): void {
        // … comment omitted
        
        // TODO: Add proper initialization logic
        // … comment omitted
    }
}

/**
 * Utility function with comprehensive documentation.
 * 
 * @param input The input string to process
 * @returns Processed string result
 */
export function processString(input: string): string {
    // … comment omitted
    if (!input || typeof input !== 'string') {
        return '';  // … comment omitted
    }
    
    // … comment omitted (5 lines)
    const trimmed = input.trim();
    const lowercase = trimmed.toLowerCase();
    const cleaned = lowercase.replace(/[^a-z0-9\s]/g, '');
    
    return cleaned;  // … comment omitted
}

// … comment omitted
function undocumentedHelper(): void {
    // … comment omitted
    const data = 'helper data';
    
    // … comment omitted
    console.log(data);  // … comment omitted
}

// … comment omitted
type ValidationResult = {
    isValid: boolean;     // … comment omitted
    errors: string[];     // … comment omitted
};

type ServiceConfig = {
    // … comment omitted
    timeout: number;      // … comment omitted
    retries: number;      // … comment omitted
    baseUrl: string;      // … comment omitted
};

// … comment omitted
interface Logger {
    info(message: string, data?: any): void;    // … comment omitted
    error(message: string, data?: any): void;   // … comment omitted
    warn(message: string, data?: any): void;    // … comment omitted
}

// … comment omitted
class ValidationError extends Error {
    constructor(public errors: string[]) {  // … comment omitted
        super(`Validation failed: ${errors.join(', ')}`);
    }
}

// … comment omitted (4 lines)
export const DEFAULT_CONFIG: ServiceConfig = {
    timeout: 5000,    // … comment omitted
    retries: 3,       // … comment omitted
    baseUrl: 'http://localhost:3000'  // … comment omitted
};
